File: src/test/java/funding/startreum/domain/virtualaccount/service/AccountPaymentServiceTest.java
package funding.startreum.domain.virtualaccount.service;

import funding.startreum.domain.funding.entity.Funding;
import funding.startreum.domain.funding.service.FundingService;
import funding.startreum.domain.project.entity.Project;
import funding.startreum.domain.project.service.ProjectService;
import funding.startreum.domain.transaction.entity.Transaction;
import funding.startreum.domain.transaction.service.TransactionService;
import funding.startreum.domain.virtualaccount.dto.request.AccountPaymentRequest;
import funding.startreum.domain.virtualaccount.dto.response.AccountPaymentResponse;
import funding.startreum.domain.virtualaccount.entity.VirtualAccount;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.bean.override.mockito.MockitoBean;

import java.math.BigDecimal;
import java.time.LocalDateTime;

import static funding.startreum.domain.transaction.entity.Transaction.TransactionType.REMITTANCE;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.when;

@SpringBootTest
class AccountPaymentServiceTest {

    @MockitoBean
    private TransactionService transactionService;

    @MockitoBean
    private ProjectService projectService;

    @MockitoBean
    private FundingService fundingService;

    @MockitoBean
    private AccountQueryService accountQueryService;

    @Autowired
    private AccountPaymentService accountPaymentService;

    @Nested
    @DisplayName("paymentByAccountId() 테스트")
    class PaymentByAccountIdTests {
        @Test
        void testPaymentByAccountId() {
            int accountId = 1;
            int projectId = 100;
            String username = "payer";
            BigDecimal paymentAmount = BigDecimal.valueOf(50);
            AccountPaymentRequest request = new AccountPaymentRequest(projectId, paymentAmount);

            // 프로젝트 객체 설정
            Project project = new Project();
            project.setProjectId(projectId);
            project.setCurrentFunding(BigDecimal.ZERO);
            when(projectService.getProject(projectId)).thenReturn(project);

            // 결제자 계좌 설정 (충전 전 잔액 200)
            VirtualAccount payerAccount = new VirtualAccount();
            payerAccount.setAccountId(accountId);
            payerAccount.setBalance(BigDecimal.valueOf(200));
            when(accountQueryService.getAccount(accountId)).thenReturn(payerAccount);

            // 프로젝트 수혜자 계좌 설정 (초기 잔액 100)
            VirtualAccount projectAccount = new VirtualAccount();
            projectAccount.setAccountId(2);
            projectAccount.setBalance(BigDecimal.valueOf(100));
            when(accountQueryService.getAccountByProjectId(projectId)).thenReturn(projectAccount);

            // 펀딩 생성 모의
            Funding funding = new Funding();
            funding.setFundingId(10);
            when(fundingService.createFunding(eq(project), eq(username), eq(paymentAmount))).thenReturn(funding);

            // 거래 생성 모의
            Transaction transaction = new Transaction();
            transaction.setTransactionId(5);
            LocalDateTime now = LocalDateTime.now();
            transaction.setTransactionDate(now);
            when(transactionService.createTransaction(eq(funding), eq(payerAccount), eq(projectAccount), eq(paymentAmount), eq(REMITTANCE)))
                    .thenReturn(transaction);

            AccountPaymentResponse response = accountPaymentService.paymentByAccountId(accountId, request, username);

            // 내부 로직: payerAccount.transferTo(paymentAmount, projectAccount)
            // 테스트를 위해 실제 transferTo가 호출되었다고 가정(예: 200-50=150)
            assertEquals(BigDecimal.valueOf(150), payerAccount.getBalance(), "결제 후 결제자 계좌 잔액이 갱신되어야 합니다.");
            // 프로젝트 funding은 프로젝트 객체 내부에서 currentfunding이 paymentAmount만큼 증가되어야 함
            assertEquals(paymentAmount, project.getCurrentFunding(), "프로젝트 currentfunding이 갱신되어야 합니다.");

            // 응답 검증
            assertEquals(transaction.getTransactionId(), response.transactionId());
            assertEquals(accountId, response.accountId());
            assertEquals(BigDecimal.valueOf(200), response.beforeMoney());
            assertEquals(paymentAmount, response.chargeAmount());
            assertEquals(payerAccount.getBalance(), response.afterMoney());
            assertEquals(now, response.transactionDate());
        }
    }

    @Nested
    @DisplayName("paymentByUsername() 테스트")
    class PaymentByUsernameTests {
        @Test
        void testPaymentByUsername() {
            int projectId = 200;
            String username = "payerUser";
            BigDecimal paymentAmount = BigDecimal.valueOf(80);
            AccountPaymentRequest request = new AccountPaymentRequest(projectId, paymentAmount);

            // 프로젝트 객체 설정
            Project project = new Project();
            project.setProjectId(projectId);
            project.setCurrentFunding(BigDecimal.ZERO);
            when(projectService.getProject(projectId)).thenReturn(project);

            // 결제자 계좌 설정 (충전 전 잔액 300)
            VirtualAccount payerAccount = new VirtualAccount();
            payerAccount.setAccountId(3);
            payerAccount.setBalance(BigDecimal.valueOf(300));
            when(accountQueryService.getAccount(username)).thenReturn(payerAccount);

            // 프로젝트 수혜자 계좌 설정 (초기 잔액 50)
            VirtualAccount projectAccount = new VirtualAccount();
            projectAccount.setAccountId(4);
            projectAccount.setBalance(BigDecimal.valueOf(50));
            when(accountQueryService.getAccountByProjectId(projectId)).thenReturn(projectAccount);

            // 펀딩 생성 모의
            Funding funding = new Funding();
            funding.setFundingId(20);
            when(fundingService.createFunding(eq(project), eq(username), eq(paymentAmount))).thenReturn(funding);

            // 거래 생성 모의
            Transaction transaction = new Transaction();
            transaction.setTransactionId(6);
            LocalDateTime now = LocalDateTime.now();
            transaction.setTransactionDate(now);
            when(transactionService.createTransaction(eq(funding), eq(payerAccount), eq(projectAccount), eq(paymentAmount), eq(REMITTANCE)))
                    .thenReturn(transaction);

            AccountPaymentResponse response = accountPaymentService.paymentByUsername(request, username);

            // 내부 로직에 따라 payerAccount.transferTo(paymentAmount, projectAccount) 실행 후: 300-80 = 220
            assertEquals(BigDecimal.valueOf(220), payerAccount.getBalance(), "결제 후 결제자 계좌 잔액이 갱신되어야 합니다.");
            // 프로젝트 currentfunding 업데이트 확인
            assertEquals(paymentAmount, project.getCurrentFunding(), "프로젝트 currentfunding이 갱신되어야 합니다.");

            assertEquals(transaction.getTransactionId(), response.transactionId());
            assertEquals(payerAccount.getAccountId(), response.accountId());
            assertEquals(BigDecimal.valueOf(300), response.beforeMoney());
            assertEquals(paymentAmount, response.chargeAmount());
            assertEquals(payerAccount.getBalance(), response.afterMoney());
            assertEquals(now, response.transactionDate());
        }
    }
}

File: src/test/java/funding/startreum/domain/virtualaccount/service/AccountPaymentServiceTest.kt
package funding.startreum.domain.virtualaccount.service

import funding.startreum.domain.funding.entity.Funding
import funding.startreum.domain.funding.service.FundingService
import funding.startreum.domain.project.entity.Project
import funding.startreum.domain.project.service.ProjectService
import funding.startreum.domain.transaction.entity.Transaction
import funding.startreum.domain.transaction.entity.Transaction.TransactionType
import funding.startreum.domain.transaction.service.TransactionService
import funding.startreum.domain.virtualaccount.dto.request.AccountPaymentRequest
import funding.startreum.domain.virtualaccount.entity.VirtualAccount
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.BDDMockito.given
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import java.math.BigDecimal
import java.time.LocalDateTime

@ExtendWith(MockitoExtension::class)
internal class AccountPaymentServiceTest {

    @Mock
    lateinit var transactionService: TransactionService

    @Mock
    lateinit var projectService: ProjectService

    @Mock
    lateinit var fundingService: FundingService

    @Mock
    lateinit var accountQueryService: AccountQueryService

    @InjectMocks
    lateinit var accountPaymentService: AccountPaymentService

    lateinit var now: LocalDateTime
    lateinit var testTransaction: Transaction

    @BeforeEach
    fun setUp() {
        now = LocalDateTime.now()
        testTransaction = Transaction().apply {
            transactionId = 999
            transactionDate = now
        }
    }

    private fun createProject(projectId: Int): Project =
        Project().apply {
            this.projectId = projectId
            currentFunding = BigDecimal.ZERO
        }

    private fun createVirtualAccount(accountId: Int, balance: BigDecimal): VirtualAccount =
        VirtualAccount().apply {
            this.accountId = accountId
            this.balance = balance
        }

    @Nested
    @DisplayName("paymentByAccountId() 테스트")
    inner class PaymentByAccountIdTests {
        @Test
        @DisplayName("정상 결제 시, 계좌 잔액/프로젝트 펀딩/거래내역 검증")
        fun testPaymentByAccountId() {
            // Given
            val accountId = 1
            val projectId = 100
            val username = "payer"
            val paymentAmount = BigDecimal.valueOf(50)
            val request = AccountPaymentRequest(projectId, paymentAmount)

            // 생성 객체
            val project = createProject(projectId)
            val payerAccount = createVirtualAccount(accountId, BigDecimal.valueOf(200))
            val projectAccount = createVirtualAccount(2, BigDecimal.valueOf(100))
            val funding = Funding().apply { fundingId = 10 }

            // Stubbing
            given(projectService.getProject(projectId)).willReturn(project)
            given(accountQueryService.getAccountByAccountId(accountId)).willReturn(payerAccount)
            given(accountQueryService.getAccountByProjectId(projectId)).willReturn(projectAccount)
            given(fundingService.createFunding(project, username, paymentAmount)).willReturn(funding)
            given(
                transactionService.createTransaction(
                    funding,
                    payerAccount,
                    projectAccount,
                    paymentAmount,
                    TransactionType.REMITTANCE
                )
            ).willReturn(testTransaction)

            // When
            val response = accountPaymentService.paymentByAccountId(accountId, request, username)

            // Then
            // 잔액 업데이트: 결제 후 결제자 200 - 50 = 150, 수혜자 100 + 50 = 150
            assertThat(payerAccount.balance).isEqualTo(BigDecimal.valueOf(150))
            assertThat(projectAccount.balance).isEqualTo(BigDecimal.valueOf(150))
            // 프로젝트의 currentFunding: 기존 0 + 50 = 50
            assertThat(project.currentFunding).isEqualTo(paymentAmount)
            // 거래 응답 검증
            assertThat(response.transactionId).isEqualTo(testTransaction.transactionId)
            assertThat(response.accountId).isEqualTo(accountId)
            assertThat(response.beforeMoney).isEqualTo(BigDecimal.valueOf(200))
            assertThat(response.chargeAmount).isEqualTo(paymentAmount)
            assertThat(response.afterMoney).isEqualTo(payerAccount.balance)
            assertThat(response.transactionDate).isEqualTo(now)
        }
    }

    @Nested
    @DisplayName("paymentByUsername() 테스트")
    inner class PaymentByUsernameTests {
        @Test
        @DisplayName("정상 결제 시, 계좌 잔액/프로젝트 펀딩/거래내역 검증")
        fun testPaymentByUsername() {
            // Given
            val projectId = 200
            val username = "payerUser"
            val paymentAmount = BigDecimal.valueOf(80)
            val request = AccountPaymentRequest(projectId, paymentAmount)

            // 생성 객체
            val project = createProject(projectId)
            val payerAccount = createVirtualAccount(3, BigDecimal.valueOf(300))
            val projectAccount = createVirtualAccount(4, BigDecimal.valueOf(50))
            val funding = Funding().apply { fundingId = 20 }

            // Stubbing
            given(projectService.getProject(projectId)).willReturn(project)
            given(accountQueryService.getAccountByUsername(username)).willReturn(payerAccount)
            given(accountQueryService.getAccountByProjectId(projectId)).willReturn(projectAccount)
            given(fundingService.createFunding(project, username, paymentAmount)).willReturn(funding)
            given(
                transactionService.createTransaction(
                    funding,
                    payerAccount,
                    projectAccount,
                    paymentAmount,
                    TransactionType.REMITTANCE
                )
            ).willReturn(testTransaction)

            // When
            val response = accountPaymentService.paymentByUsername(request, username)

            // Then
            // 잔액 업데이트: 결제 후 결제자 300 - 80 = 220, 수혜자 50 + 80 = 130
            assertThat(payerAccount.balance).isEqualTo(BigDecimal.valueOf(220))
            assertThat(projectAccount.balance).isEqualTo(BigDecimal.valueOf(130))
            // 프로젝트의 currentFunding: 기존 0 + 80 = 80
            assertThat(project.currentFunding).isEqualTo(paymentAmount)
            // 거래 응답 검증
            assertThat(response.transactionId).isEqualTo(testTransaction.transactionId)
            assertThat(response.accountId).isEqualTo(payerAccount.accountId)
            assertThat(response.beforeMoney).isEqualTo(BigDecimal.valueOf(300))
            assertThat(response.chargeAmount).isEqualTo(paymentAmount)
            assertThat(response.afterMoney).isEqualTo(BigDecimal.valueOf(220))
            assertThat(response.transactionDate).isEqualTo(now)
        }
    }
}